landoapi/api/treestatus.py

```diff
@@ -0,0 +1,609 @@
+    """Handler for `DELETE /stack/{id}`."""
+    if revert not in {0, 1, None}:
+        raise ProblemException(
+            400,
+            "Unexpected value for `revert`.",
+            "Unexpected value for `revert`.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
+
+    return revert_change(id, revert=bool(revert))
+
+
+@auth.require_auth0(
+    groups=(auth.TREESTATUS_USERS), scopes=("lando", "profile", "email"), userinfo=True
+)
+def discard_change(id: int):
+    """Handler for `DELETE /stack2/discard/{id}`.
+
+    Remove the change from the stack without changing the settings on the tree.
+    """
+    return revert_change(id, revert=False)
+
+
+@auth.require_auth0(
+    groups=(auth.TREESTATUS_USERS), scopes=("lando", "profile", "email"), userinfo=True
+)
+def restore_change(id: int):
+    """Handler for `DELETE /stack2/restore/{id}`.
+
+    Applies the settings that were present before the change to the affected trees.
+    """
+    return revert_change(id, revert=True)
+
+
+@result_object_wrap
+def get_trees() -> dict:
+    """Handler for `GET /trees`."""
+    return {tree.tree: combinedtree_as_dict(tree) for tree in get_combined_trees()}
+
+
+@auth.require_auth0(
+    groups=(auth.TREESTATUS_USERS), scopes=("lando", "profile", "email"), userinfo=True
+)
+def update_trees(body: dict):
+    """Handler for `PATCH /trees`."""
+    session = db.session
+
+    # Fetch all trees.
+    trees = get_combined_trees(body["trees"])
+
+    # Check that we fetched all the trees.
+    if len(trees) != len(body["trees"]):
+        trees_diff = set(body["trees"]) - set(tree.tree for tree in trees)
+        raise ProblemException(
+            404,
+            f"Could not fetch the following trees: {trees_diff}"
+            "Could not fetch all the requested trees.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+
+    # Check for other constraints.
+    if "tags" not in body and body.get("status") == "closed":
+        raise ProblemException(
+            400,
+            "Tags are required when closing a tree.",
+            "Tags are required when closing a tree.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
+
+    if (
+        "remember" in body
+        and body["remember"] is True
+        and any(field not in body for field in {"status", "reason", "tags"})
+    ):
+        raise ProblemException(
+            400,
+            "Must specify status, reason and tags to remember the change.",
+            "Must specify status, reason and tags to remember the change.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
+
+    # Update the trees as requested.
+    new_status = body.get("status")
+    new_reason = body.get("reason")
+    new_motd = body.get("message_of_the_day")
+    new_tags = body.get("tags", [])
```

Comment:
`status`, `reason`, `message_of_the_day`, and `tags` are repeated quite a bit in this method. I think if this is refactored to reuse these keys in a more generic way, it would make this a bit easier to read. E.g.,
```suggestion
    keys = ("status", "reason", "tags")
    new_values = {k: body.get(k) for k in keys + ("message_of_the_day",)}
```